src/vector_windows.rs

```diff
@@ -0,0 +1,222 @@
+        // TODO: It is currently not possible to change the description of the service.
+        // Waiting for the following PR to get merged in
+        // https://github.com/mullvad/windows-service-rs/pull/32
+        //
+        // service.set_description(&self.description);
+        Ok(())
+    }
+
+    fn uninstall_service(service: &ServiceDefinition) -> Result<()> {
+        let service_access =
+            ServiceAccess::QUERY_STATUS | ServiceAccess::STOP | ServiceAccess::DELETE;
+        let service = open_service(&service, service_access)?;
+
+        let service_status = service.query_status()?;
+        if service_status.current_state != ServiceState::Stopped {
+            service.stop()?;
+            thread::sleep(Duration::from_secs(1));
+        }
+
+        service.delete()?;
+        Ok(())
+    }
+
+    pub fn open_service(
+        service: &ServiceDefinition,
+        access: windows_service::service::ServiceAccess,
+    ) -> Result<windows_service::service::Service> {
+        let manager_access = ServiceManagerAccess::CONNECT;
+        let service_manager = ServiceManager::local_computer(None::<&str>, manager_access)?;
+
+        let service = service_manager.open_service(&service.name, access)?;
+        Ok(service)
+    }
+}
+
+define_windows_service!(ffi_service_main, win_main);
+
+fn win_main(arguments: Vec<OsString>) {
+    if let Err(_e) = run_service(arguments) {}
+}
+
+pub fn run() -> Result<()> {
+    service_dispatcher::start(SERVICE_NAME, ffi_service_main)
+}
+
+fn run_service(_arguments: Vec<OsString>) -> Result<()> {
+    let (shutdown_tx, shutdown_rx) = mpsc::channel();
+
+    let event_handler = move |control_event| -> ServiceControlHandlerResult {
+        match control_event {
+            // Notifies a service to report its current status information to the service
+            // control manager. Always return NoError even if not implemented.
+            ServiceControl::Interrogate => ServiceControlHandlerResult::NoError,
+
+            // Handle stop
+            ServiceControl::Stop => {
+                shutdown_tx.send(()).unwrap();
+                ServiceControlHandlerResult::NoError
+            }
+
+            _ => ServiceControlHandlerResult::NotImplemented,
+        }
+    };
+    let status_handle =
+        windows_service::service_control_handler::register(SERVICE_NAME, event_handler)?;
+
+    let code = match start_service() {
+        Ok(vector) => {
+            status_handle.set_service_status(ServiceStatus {
+                service_type: SERVICE_TYPE,
+                current_state: ServiceState::Running,
+                controls_accepted: ServiceControlAccept::STOP,
+                exit_code: ServiceExitCode::Win32(0),
+                checkpoint: 0,
+                wait_hint: Duration::default(),
+            })?;
+
+            let topology = vector.topology;
+
+            shutdown_rx.recv();
+            let shutdown = topology.stop();
```

Comment:
I guess I need to wait for the topology to stop here. Should I use the same code pattern than `main.rs` and `await` on the `Future` inside a  `Runtime::block_on` ? Ideally, we also need some kind of timeout and force the server to shut down.

Reply:
Will Windows not issue something like a SIGKILL at a certain timeout? Whoa.

Reply:
Windows does not have unix signals. An application can use the WinAPI [SetConsoleCtrlHandler](https://docs.microsoft.com/en-us/windows/console/setconsolectrlhandler) for events like Ctrl-C but the system won't emit any signal.

I'm not sure what the SCM does when the timeout is reached though.

Reply:
Hmmm okay. I think it might be worthwhile for us to use a timeout yes. So I guess the same strategy main.rs uses then is best.

Reply:
Fixed in https://github.com/timberio/vector/pull/2896/commits/79b00d20a610022561fe7940026bbb13ea688901

The future returned by `RunningTopology::stop()` is now correctly await-ed. It already has an internal deadline timer, so I did not add one.